The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12273458894/job/34244288043) [(plain)](https://github.com/rust-lang/rust/commit/6e867e1b0bd1b8c41faaaa8cf16342fbaff22cc7/checks/34244288043/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.856 Building wheels for collected packages: reuse
#13 2.857   Building wheel for reuse (pyproject.toml): started
#13 3.103   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.104   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.104   Stored in directory: /tmp/pip-ephem-wheel-cache-b7sb97dp/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.107 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.513 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.513 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.051 Collecting virtualenv
#13 4.051 Collecting virtualenv
#13 4.088   Downloading virtualenv-20.28.0-py3-none-any.whl (4.3 MB)
#13 4.169      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 54.4 MB/s eta 0:00:00
#13 4.208 Collecting distlib<1,>=0.3.7
#13 4.212   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#13 4.220      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 77.4 MB/s eta 0:00:00
#13 4.251 Collecting platformdirs<5,>=3.9.1
#13 4.254   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.290 Collecting filelock<4,>=3.12.2
#13 4.293   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.374 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.558 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.28.0
#13 DONE 4.6s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      319424 kB
DirectMap2M:     7020544 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2024-11-27/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
fmt check
Diff in /checkout/compiler/rustc_target/src/spec/targets/wasm32_unknown_emscripten.rs:32:
             std: Some(true),
         },
         pointer_width: 32,
-        data_layout: "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-i128:128-f128:64-n32:64-S128-ni:1:10:20".into(),
+        data_layout: "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-i128:128-f128:64-n32:64-S128-ni:1:10:20"
+            .into(),
         arch: "wasm32".into(),
     }
     }
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_target/src/spec/targets/thumbv8m_base_none_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/wasm32_unknown_emscripten.rs" "/checkout/tests/mir-opt/address_of.rs" "/checkout/tests/mir-opt/issue_120925_unsafefncast.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_wrs_vxworks.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv64gc_unknown_freebsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_pc_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv4t_none_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_pc_windows_gnullvm.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7a_none_eabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32im_unknown_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/i586_pc_nto_qnx700.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsisa32r6_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc_wrs_vxworks_spe.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7neon_linux_androideabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32imc_esp_espidf.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv4t_none_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_apple_ios_macabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv64gc_unknown_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv64gc_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_apple_tvos.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_pc_solaris.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_apple_ios.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_unknown_trusty.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_hurd_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32imafc_unknown_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv64imac_unknown_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsel_unknown_none.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_apple_watchos_sim.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv5te_unknown_linux_uclibceabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_hermit.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_unknown_linux_musleabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_trusty.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32imafc_esp_espidf.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc64_unknown_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsisa32r6el_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv6_unknown_freebsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32ima_unknown_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsisa64r6el_unknown_linux_gnuabi64.rs" "/checkout/compiler/rustc_target/src/spec/targets/arm_unknown_linux_musleabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/arm64e_apple_darwin.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_be_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsel_unknown_linux_uclibc.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32imac_unknown_nuttx_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/sparc64_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_uwp_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/armeb_unknown_linux_gnueabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/loongarch64_unknown_none.rs" "/checkout/compiler/rustc_target/src/spec/targets/mips64_unknown_linux_gnuabi64.rs" "/checkout/compiler/rustc_target/src/spec/targets/wasm32_wasip2.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7neon_unknown_linux_gnueabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsisa64r6_unknown_linux_gnuabi64.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv5te_unknown_linux_gnueabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv8m_base_nuttx_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/xtensa_esp32_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_unknown_linux_gnueabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/sparc_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc64_unknown_freebsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/mips_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc_wrs_vxworks.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv64_linux_android.rs" "/checkout/compiler/rustc_target/src/spec/targets/arm64e_apple_tvos.rs" "/checkout/tests/mir-opt/issue_38669.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Wed Dec 11 09:49:06 UTC 2024
  network time: Wed, 11 Dec 2024 09:49:07 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/compiler/rustc_codegen_llvm/src/context.rs
+++ b/compiler/rustc_codegen_llvm/src/context.rs
@@ -164,6 +164,11 @@ pub(crate) unsafe fn create_module<'ll>(
             // See https://github.com/llvm/llvm-project/pull/118004
             target_data_layout = target_data_layout.replace("-i128:128", "");
         }
+        if sess.target.arch.starts_with("wasm32") || sess.target.arch.starts_with("wasm64") {
+            // LLVM 20 updates the wasm(32|64) layout to correctly align 128 bit integers to 128 bit.
+            // See https://github.com/llvm/llvm-project/pull/119204
+            target_data_layout = target_data_layout.replace("-i128:128", "");
+        }
     }
 
     // Ensure the data-layout values hardcoded remain the defaults.

--- a/compiler/rustc_target/src/spec/targets/wasm32_unknown_emscripten.rs
+++ b/compiler/rustc_target/src/spec/targets/wasm32_unknown_emscripten.rs
@@ -32,7 +32,8 @@ pub(crate) fn target() -> Target {
             std: Some(true),
         },
         pointer_width: 32,
-        data_layout: "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-f128:64-n32:64-S128-ni:1:10:20".into(),
+        data_layout: "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-i128:128-f128:64-n32:64-S128-ni:1:10:20"
+            .into(),
         arch: "wasm32".into(),
         options: opts,
     }

--- a/compiler/rustc_target/src/spec/targets/wasm32_unknown_unknown.rs
+++ b/compiler/rustc_target/src/spec/targets/wasm32_unknown_unknown.rs
@@ -37,7 +37,7 @@ pub(crate) fn target() -> Target {
             std: Some(true),
         },
         pointer_width: 32,
-        data_layout: "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-n32:64-S128-ni:1:10:20".into(),
+        data_layout: "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-i128:128-n32:64-S128-ni:1:10:20".into(),
         arch: "wasm32".into(),
         options,
     }

--- a/compiler/rustc_target/src/spec/targets/wasm32_wasip1.rs
+++ b/compiler/rustc_target/src/spec/targets/wasm32_wasip1.rs
@@ -55,7 +55,7 @@ pub(crate) fn target() -> Target {
             std: Some(true),
         },
         pointer_width: 32,
-        data_layout: "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-n32:64-S128-ni:1:10:20".into(),
+        data_layout: "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-i128:128-n32:64-S128-ni:1:10:20".into(),
         arch: "wasm32".into(),
         options,
     }

--- a/compiler/rustc_target/src/spec/targets/wasm32_wasip1_threads.rs
+++ b/compiler/rustc_target/src/spec/targets/wasm32_wasip1_threads.rs
@@ -66,7 +66,7 @@ pub(crate) fn target() -> Target {
             std: Some(true),
         },
         pointer_width: 32,
-        data_layout: "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-n32:64-S128-ni:1:10:20".into(),
+        data_layout: "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-i128:128-n32:64-S128-ni:1:10:20".into(),
         arch: "wasm32".into(),
         options,
     }

--- a/compiler/rustc_target/src/spec/targets/wasm32_wasip2.rs
+++ b/compiler/rustc_target/src/spec/targets/wasm32_wasip2.rs
@@ -66,7 +66,7 @@ pub(crate) fn target() -> Target {
             std: Some(true),
         },
         pointer_width: 32,
-        data_layout: "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-n32:64-S128-ni:1:10:20".into(),
+        data_layout: "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-i128:128-n32:64-S128-ni:1:10:20".into(),
         arch: "wasm32".into(),
         options,
     }

--- a/compiler/rustc_target/src/spec/targets/wasm32v1_none.rs
+++ b/compiler/rustc_target/src/spec/targets/wasm32v1_none.rs
@@ -44,7 +44,7 @@ pub(crate) fn target() -> Target {
             std: Some(false),
         },
         pointer_width: 32,
-        data_layout: "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-n32:64-S128-ni:1:10:20".into(),
+        data_layout: "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-i128:128-n32:64-S128-ni:1:10:20".into(),
         arch: "wasm32".into(),
         options,
     }

--- a/compiler/rustc_target/src/spec/targets/wasm64_unknown_unknown.rs
+++ b/compiler/rustc_target/src/spec/targets/wasm64_unknown_unknown.rs
@@ -40,7 +40,7 @@ pub(crate) fn target() -> Target {
             std: None, // ?
         },
         pointer_width: 64,
-        data_layout: "e-m:e-p:64:64-p10:8:8-p20:8:8-i64:64-n32:64-S128-ni:1:10:20".into(),
+        data_layout: "e-m:e-p:64:64-p10:8:8-p20:8:8-i64:64-i128:128-n32:64-S128-ni:1:10:20".into(),
         arch: "wasm64".into(),
         options,
     }

